kythe/java/com/google/devtools/kythe/extractors/java/bazel/aspect.bzl

```diff
@@ -0,0 +1,136 @@
+# Copyright 2019 The Kythe Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+load("//tools/build_rules/verifier_test:verifier_test.bzl", "extract")
+
+_mnemonic = "Javac"
+
+def _extract_java(target, ctx):
+    if JavaInfo not in target or not hasattr(ctx.rule.attr, "srcs"):
+        return None
+
+    javac_action = None
+    for a in target.actions:
+        if a.mnemonic == _mnemonic:
+            javac_action = a
+            break
+
+    if not javac_action:
+        return None
+
+    kzip = ctx.actions.declare_file(ctx.label.name + ".xa.java.kzip")
+
+    info = target[JavaInfo]
+    compilation = info.compilation_info
+    annotations = info.annotation_processing
+
+    source_files = []
+    for src in ctx.rule.files.srcs:
+        source_files += [src.path]
+
+    classpath = [j.path for j in compilation.compilation_classpath.to_list()]
+    bootclasspath = [j.path for j in compilation.boot_classpath]
+
+    processorpath = []
+    processors = []
+    if annotations and annotations.enabled:
+        processorpath += [j.path for j in annotations.processor_classpath.to_list()]
+        processors = annotations.processor_classnames
+
+    output_jar = []
+    for jar in info.outputs.jars:
+        output_jar += [jar.class_jar.path]
+
+    # TODO(schroederc): sourcegendir
+    xa = struct(**{
+        "owner": str(target.label),
+        "mnemonic": _mnemonic,
+        "[blaze.JavaCompileInfo.java_compile_info]": struct(
+            outputjar = output_jar,
+            classpath = classpath,
+            source_file = source_files,
+            javac_opt = compilation.javac_options,
+            processor = processors,
+            processorpath = processorpath,
+            bootclasspath = bootclasspath,
+        ),
+    })
+    textXA = ctx.actions.declare_file(ctx.label.name + ".xa.textproto")
```

Comment:
nit: I think Starlark is typically lower_with_unders for local variables? (This is what's recommended by the internal guide, at least).